### ISSUE TITLE
Start script hangs without error when there are no available inotify watches

### ISSUE THREAD

**Issue by trygveaa** [react-community/create-react-native-app#234]
## Description

I tried to run `npm start` in a new project (i.e. I ran the four commands under "Quick Overview"), and it hanged indefinitely (well, at least for several minutes until I aborted) after the output "Starting packager...".

Running `react-native start` gave the error `ERROR  watch {path} ENOSPC`, which made me realize that there were no available inotify watches.

After increasing `fs.inotify.max_user_watches`, `npm start` started and ran normally.

## Expected Behavior

`npm start` should give a warning or exit with an error saying that there are no available watches, so it can't start.

## Observed Behavior

`npm start` hanged with this output, and never continued:

```
$ npm start

> my-app@0.1.0 start /home/trygve/dev/create-react-native-app/my-app
> react-native-scripts start

13.49.26: Starting packager...
```

## Environment

Please run these commands in the project folder and fill in their results:

* `npm ls react-native-scripts`: react-native-scripts@0.0.30
* `npm ls react-native`: react-native@0.44.2
* `npm ls expo`: expo@17.0.0
* `node -v`: v7.10.0
* `npm -v`: 4.6.1
* `yarn --version`: 0.24.6
* `watchman version`: Not installed

Also specify:

1. Operating system: Arch Linux
2. Phone/emulator/simulator & version: Not relevant

## Reproducible Demo

Ran `npm start` in a clean project just created, so nothing to provide here.

**Comment by brentvatne** [comment]
It seems pretty reasonable to handle this error with a useful message as you described, as well as, of course, to not hang but to instead show some useful information and exit instead. If anyone is interested in contributing, this is a useful place to look: https://github.com/react-community/create-react-native-app/blob/master/react-native-scripts/src/util/packager.js#L55

**Comment by nimir** [comment]
I've the start script also hanging without error but seems for a different reason as i've already increased my inotify watchers and running react-native start works fine without a problem!

**Environment:**

- npm ls react-native-scripts: react-native-scripts@0.0.31
- npm ls react-native: react-native@0.44.3
- npm ls expo: expo@17.0.0
- node -v: v6.9.5
- npm -v: 5.0.3
- yarn --version: Not installed
- watchman version: 4.7.0

Also tried:
- `rm -rf node_modules` then fresh npm install
- `watchman watch-del-all`
- `npm start --clear-cache`

How can i debug this to find what's causing the issue? tried to look at the packager.js and scripts/start.js for verbose flags without luck!

**Comment by thinq4yourself** [comment]
👍 on this 20 day old issue. First go at CRNA for our new app with latest and this happens to use, too.

**Comment by k3a** [comment]
Confirming the issue, seen it too and it's pretty bad bug because it affects newcomers - I've created a new empty React native project and it simply hang on "Starting packager...". Confirming increase of that sysctl fixed the issue.

@nimir : try this in your project root if not tried yet:  ./.bin/react-native start

**Comment by s417-lama** [comment]
I had the same issue on Ubuntu 16.04.
Try to install watchman.
If you get some error about watchman when running `npm start`, see: https://github.com/react-community/create-react-native-app/issues/51

**Comment by tetralix** [comment]
@k3a Can you please provide more details how "increase of that sysctl" fixed this issue?

**Comment by k3a** [comment]
@tetralix On Linux do "sudo sysctl -w fs.inotify.max_user_watches=10000". This changes kernel sysctl value. This is temporary until reboot. To make it permanent, add fs.inotify.max_user_watches=10000 to a file in /etc/sysctl.d directory or /etc/sysctl.conf depending on distro.

**Comment by tetralix** [comment]
@k3a I would never figure out that problem is related to user watches parameter. Worked for me, but increased to 12k. OS: Fedora 25

**Comment by luffs** [comment]
Spent hours googling, reinstalling, upgrading, downgrading, switching to yarn until I found this issue. Thank you!

+1 to have a error message...

Used this guide to increase the inotify watchers max:
https://github.com/guard/listen/wiki/Increasing-the-amount-of-inotify-watchers

**Comment by penniath** [comment]
Thanks @k3a, you saved my day!

**Comment by akshaynaik404** [comment]
I had the same issue on Ubuntu 16.04.2 LTS. As @s417-lama mentioned #51 contains the solution that worked for me.

But this bug is very frustrating for **beginners** and should be given priority. @dikaiosune has already taken a note of this. I hope this gets *fixed soon*.

**Comment by ashishrvtech** [comment]
@k3a It works thanks!!!

**Comment by darkguy2008** [comment]
I also have the same issue, using OSX Sierra 10.12.5. sysctl returns ```sysctl: unknown oid 'fs.inotify.max_user_watches'```. Any ideas? :( I can't start up any react iOS app, the simulator does, but nothing else happens and it gets stuck in the Packager thing...

**Comment by vs1682** [comment]
@brentvatne @dikaiosune We should also add this in [Getting Started](https://github.com/react-community/create-react-native-app#getting-started) section. So that it will easier for beginner.

**Comment by kevz93** [comment]
The command should be part of the start script.

**Comment by k3a** [comment]
@darkguy2008 `fs.inotify.max_user_watches` is Linux-specific sysctl and I mainly use Linux. Mac OS implementation probably uses FSEvents and I don't know if there are any limits. Maybe your problem is somewhere else. You can try debugging it by seeing all system calls. Try `dtruss npm start` and look for suspicious error return codes.

**Comment by darkguy2008** [comment]
@k3a yes I figured out after some research hehe. Although I appreciate your answer, you can see I fixed my error using those two commands, which are the equivalents for OSX, which I agree with @vs1682 they should be added to the start script :)

**Comment by vs1682** [comment]
As mentioned by @kevz93 I'm thinking of adding a part where it set the inotify(linux) and kern(macOS) watches and raising a PR.
I'm thinking of adding that part here https://github.com/react-community/create-react-native-app/blob/master/react-native-scripts/src/util/packager.js#L63.
@brentvatne @dikaiosune Please let me if it's the place to add.

**Comment by trygveaa** [comment]
I don't think it should change sysctl config automatically, at least not without asking the user first. And it would require root, which this script shouldn't have.

Rather I think it should inform the user if there isn't enough watchers, and print the commands the user should run.

**Comment by vs1682** [comment]
@trygveaa 

> I don't think it should change sysctl config automatically, at least not without asking the user first. And it would require root, which this script shouldn't have.

Right.

> Rather I think it should inform the user if there isn't enough watchers, and print the commands the user should run.

I'm actually working on this solution only.

Thanks.

**Comment by nickjuntilla** [comment]
This also happens for me on Sierra 10.12.5. A fresh create react native project works but this one I've been working on will no longer start. I've set all my package.json files back to to how they were with a fresh project, but it's just not working.

**Comment by nickjuntilla** [comment]
I solved this by starting over with a new create react app.

**Comment by nickjuntilla** [comment]
Now I'm getting a very strange error:

 2017-08-12 14:50 node[4105] (FSEvents.framework) FSEventStreamStart: register_with_server: ERROR: f2d_register_rpc() => (null) (-22)

Is there a recommended node version number aside from the current stable?

**Comment by k3a** [comment]
@Nicholas https://github.com/facebook/react-native/issues/10028 It seems to be related to this issue, try increasing maxfiles (a few comments above)

**Comment by vs1682** [comment]
Raised a [PR](https://github.com/react-community/create-react-native-app/pull/358) to solve this issue.
Please check and let me know.

Note: As I don't have a mac, I was not able to check this for macOS. If would appreciate if someone can do that. Thanks.

**Comment by darkguy2008** [comment]
@vs1682 I do have one, but I don't know how to test that PR :( if you can guide me, I'll happily test for you.

**Comment by vs1682** [comment]
@darkguy2008 you need to fetch the pull request in your local repo and follow the contribution guide to set it up locally.
Make sure that your system's kern.maxfiles and kern.maxfilesperproc are set to default.

**Comment by festiveox** [comment]
This is happening to me on Unbutu, I try to run sudo sysctl -w fs.inotify.max_user_watches=10000 but it doesn't fix the issue either... any other fix gor this on unbutu? Or anyone else experiencing this?

**Comment by trygveaa** [comment]
@festiveox: 10000 might not be enough. Try to increase it more.

**Comment by festiveox** [comment]
@trygveaa thank you very much I increased it to 120000 and now it's working.

**Comment by bitmega** [comment]
@darkguy2008 On my Mac, I have to change the value a little bit
```
  sudo sysctl -w kern.maxfiles=5242880
  sudo sysctl -w kern.maxfilesperproc=524288
```

**Comment by joevallender** [comment]
Required 10485760 for me

**Comment by mikeaustin** [comment]
Well, this isn't going well for me. Using create-react-native-app, I hit this:

    5:28:00 PM: Unable to start server
    See https://git.io/v5vcn for more information, either install watchman or run the following snippet:
      sudo sysctl -w kern.maxfiles=5242880
      sudo sysctl -w kern.maxfilesperproc=524288

"install watchman" is a bit brief. I did "npm install -g watchman", and then re-ran. It did something more, and I was able to use the QR code. But then I get this

    jest-haste-map: @providesModule naming collision:
      Duplicate module name: ReactNative
      Paths: /Users/mikeaustin/Projects/ReactNative/node_modules/react-native/Libraries/Renderer/shims/ReactNative.js collides with /Users/mikeaustin/Projects/ReactNative/package.json

    This warning is caused by a @providesModule declaration with the same name across two different files.

And the last line on the console, with a red background and error stats on my phone:

    5:33:13 PM: Failed building JavaScript bundle

So, this is not easy at all :) I'm familiar with React, and also Xcode but not React Native.

**Comment by baptistemanson** [comment]
Did you call your app reactNative by any chance? :)

**Comment by sathiz** [comment]
I was getting the same error as @mikeaustin 

> Unable to start server
> See https://git.io/v5vcn for more information, either install watchman or run the following snippet:
>   sudo sysctl -w kern.maxfiles=5242880
>   sudo sysctl -w kern.maxfilesperproc=524288

but once i installed watchman, it worked like a charm!

**Comment by tiongshunooi** [comment]
+1

**Comment by tiongshunooi** [comment]
+1 to have an error message...

**Comment by Meekohi** [comment]
Could anyone give an ELI5 of what's going on here? If watchman is a dependency why isn't it just included in the first place?

**Comment by dikaiosune** [comment]
@Meekohi it's...complicated. The main reasons are that watchman isn't distributed via npm, and maintaining our own npm distribution of the binaries would be rather difficult.

**Comment by Meekohi** [comment]
Thanks, actually this is what screwed me up -- there is a `watchman` on `npm` (https://github.com/dfjones/watchman) which does almost exactly the same thing, and not knowing better I didn't realize this _wasn't_ the watchman I was supposed to install (and thus my confusion above). Not sure what a good solution is here given the conflicting package names...

**Comment by lrvlr** [comment]
I'm so confused and tired over here. I had the same error as @sathiz  and @mikeaustin  
Unable to start server See https://git.io/v5vcn for more information, either install watchman or run the following snippet: sudo sysctl -w kern.maxfiles=5242880 sudo sysctl -w kern.maxfilesperproc=524288  

I installed watchman, but now it just hangs on:

> testing@0.1.0 start /.../testing
> react-native-scripts start

9:31:05 PM: Starting packager...

I can't make it work.
I've also tried using exp start, it gives me the QR but on my phone I get "failed to complete tunnel connection" and nothing happens.



    npm ls react-native-scripts: react-native-scripts@1.7.0 
    npm ls react-native: react-native@0.49.5 
    npm ls expo: expo@22.0.3 
    node -v: v6.12.0
    npm -v: 4.6.1
    yarn --version: 0.27
    watchman version: Please specify a target and action ???
    Operating system: Ubuntu 17.04

Can anyone help?

**Comment by brentvatne** [comment]
@lrvlr - can you jump onto our slack and dm me? https://slack.expo.io -- I am brent on there

**Comment by thomasyimgit** [comment]
FYI, I ran into the same issue again, but fixed with running
```
  sudo sysctl -w kern.maxfiles=5242880
  sudo sysctl -w kern.maxfilesperproc=524288
```
I think it would be weird and confusing to come across this error at very beginning. And it would not hurt to mention this in an obvious position of the document, rather than in this issue.

**Comment by willnwhite** [comment]
@anp Please could you link to watchman? I think the "install watchman" error message should include this link.

**Comment by willnwhite** [comment]
@Meekohi, please could you post the link to the right watchman here?

**Comment by Meekohi** [comment]
https://github.com/facebook/watchman

**Comment by alper** [comment]
I npm installed watchman but that did not fix this problem.

**Comment by Meekohi** [comment]
@alper you can't get the watchman off npm. Get it from Facebook's github as linked above. The npm version is now officially deprecated so you should get a warning.

**Comment by jsstrn** [comment]
To solve this issue, just install Watchman via Homebrew (for macOS) with:

```sh
$ brew update
$ brew install watchman
```

Source: [Watchman docs](https://facebook.github.io/watchman/docs/install.html#installing-on-os-x-via-homebrew). For other operating systems, refer to their [docs](https://facebook.github.io/watchman/docs/install.html).

**Comment by sendsent** [comment]
why does sudo sysctl -w kern.maxfilesperproc=524288 solve this problem?

**Comment by Kristories** [comment]
@sendsent

```js
if (process.platform !== 'win32') {
  var watchmanExists = _crossSpawn2.default.sync('which', ['watchman']).status === 0;

  if (process.platform === 'darwin' && !watchmanExists) {
    var watcherDetails = _crossSpawn2.default.sync('sysctl', ['kern.maxfiles']).stdout.toString();
    if (parseInt(watcherDetails.split(':')[1].trim()) < 5242880) { // <--------------
      _log2.default.withTimestamp(_chalk2.default.red('Unable to start server') + '\nSee https://git.io/v5vcn for more information, either install watchman or run the following snippet:\n' + _chalk2.default.cyan('  sudo sysctl -w kern.maxfiles=5242880\n  sudo sysctl -w kern.maxfilesperproc=524288') + '\n        ');
      process.exit(1);
    }
  } else if (!watchmanExists) {
    try {
      var _watcherDetails = _crossSpawn2.default.sync('sysctl', ['fs.inotify.max_user_watches']).stdout.toString();
      if (parseInt(_watcherDetails.split('=')[1].trim()) < 12288) { // <--------------
        _log2.default.withTimestamp(_chalk2.default.red('Unable to start server') + '\n  See https://git.io/v5vcn for more information, either install watchman or run the following snippet:\n  ' + _chalk2.default.cyan('  sudo sysctl -w fs.inotify.max_user_instances=1024\n    sudo sysctl -w fs.inotify.max_user_watches=12288'));
        process.exit(1);
      }
    } catch (e) {
      _log2.default.withTimestamp('Warning: Unable to run `sysctl fs.inotify.max_user_watches`. If you encounter issues, please refer to https://git.io/v5vcn');
    }
  }
}
```

**Comment by ide** [comment]
@vinniejames You should install Watchman from Homebrew per the instructions, not npm.

**Comment by vinniejames** [comment]
@ide the notice doesn't say anything about using brew `either install watchman or run the following snippet`

**Comment by ide** [comment]
C.f. various docs online but it seems the error message should probably be more precise to point people towards Homebrew.

**Comment by AndresCuervo** [comment]
Just ran into this issue. Agreed with the above commenters that something like:

```
On MacOS run:

    $ brew install watchman
```

should be added to the error message. Of course adding other relevant information for other OS's as well. Pretty bad developer experience to have to come to this GH issue & skim to find out what I needed to do just to run the example setup.

**Comment by geoffreywiseman** [comment]
Maybe just include a URL to watchman -- no sense providing install instructions as they vary by platform and could be subject to change, no sense coupling CRNA to the installation instructions for watchman IMO.

**Comment by noodlebreak** [comment]
Nothing worked for me...
```
npm ls react-native-scripts - react-native-scripts@1.14.0 

npm ls react-native - react-native@0.55.3 

npm ls expo - expo@27.0.1 

node -v - v8.11.1
npm -v - 5.6.0

yarn --version

The program 'yarn' is currently not installed. You can install it by typing:
sudo apt install cmdtest

watchman version
watchman: command not found
```

On running `npm start`:

```
 (968)-> npm start

> billi@0.1.0 start /home/ash/projects/myproject
> react-native-scripts start

01:09:56: Starting packager...

```

**Comment by DStt** [comment]
`cat /proc/sys/fs/inotify/max_user_watches
> 1874919423`

On running `npm start`

`> mobile-flashcards@0.1.0 start /home/dsa/project/test/mobile-flashcards
> react-native-scripts start
13:40:40: Warning: Unable to run `sysctl fs.inotify.max_user_watches`. If you encounter issues, please refer to https://git.io/v5vcn`

I don't receive the warning if i launch it with **sudo**
Why I have to launch a local npm app with root privileges?

**Comment by willxy** [comment]
OSX; ran into same thing, had to read the entire thread to find out what "install watchman" means and that it doesn't mean npm, agree with comments that it should at least point to the watchman website, although having it suggest "brew install watchman" would be by far the most helpful.

**Comment by kwngo** [comment]
Thanks @willxy

**Comment by anp** [comment]
@willxy you're right! Since this issue is apparently showing up in search results for lots of people, I decided to lock this and to hide all comments that don't contain important information for fixing the issue. Hopefully it should be much easier to read the history now. If after installing watchman (NOT using the `watchman` npm package) and raising inotify/fsevents limits you're still having issues, please open a new issue! Thanks.
